File: apps/web/lib/geocode.ts
/**
 * Customer-address geocoding via Mapbox.
 *
 * Bounded to the UK (gb), filtered to address/poi/place types. Returns
 * a single best candidate with a confidence band:
 *
 *   "rooftop"  → relevance ≥ 0.9 AND match_code.exact_match (or address type)
 *   "street"   → relevance 0.7–0.9 OR street type
 *   "postcode" → relevance 0.5–0.7 OR postcode-only match
 *   "failed"   → no usable match
 *
 * Customers with confidence < "street" go to a manual review queue —
 * routing them blindly would put pins in the wrong village.
 */

export type GeocodeConfidence = "rooftop" | "street" | "postcode" | "failed";

export interface GeocodeResult {
  lat: number | null;
  lng: number | null;
  confidence: GeocodeConfidence;
  matchedAddress: string | null;
  raw: unknown;
}

const MAPBOX_GEOCODE_URL =
  "https://api.mapbox.com/geocoding/v5/mapbox.places";

export async function geocodeAddress(
  address: string,
  opts: { token?: string } = {},
): Promise<GeocodeResult> {
  const token = opts.token ?? process.env.MAPBOX_TOKEN;
  if (!token) {
    return {
      lat: null,
      lng: null,
      confidence: "failed",
      matchedAddress: null,
      raw: { error: "MAPBOX_TOKEN not set" },
    };
  }
  const trimmed = address?.trim();
  if (!trimmed) {
    return { lat: null, lng: null, confidence: "failed", matchedAddress: null, raw: null };
  }

  const url = new URL(`${MAPBOX_GEOCODE_URL}/${encodeURIComponent(trimmed)}.json`);
  url.searchParams.set("access_token", token);
  url.searchParams.set("country", "gb");
  url.searchParams.set("limit", "1");
  url.searchParams.set("types", "address,poi,place,postcode");
  url.searchParams.set("autocomplete", "false");

  const res = await fetch(url, { method: "GET" });
  if (!res.ok) {
    return {
      lat: null,
      lng: null,
      confidence: "failed",
      matchedAddress: null,
      raw: { status: res.status, statusText: res.statusText },
    };
  }
  const body = await res.json();
  const feature = body?.features?.[0];
  if (!feature || !feature.center) {
    return { lat: null, lng: null, confidence: "failed", matchedAddress: null, raw: body };
  }

  const [lng, lat] = feature.center as [number, number];
  return {
    lat,
    lng,
    confidence: classifyConfidence(feature),
    matchedAddress: feature.place_name ?? null,
    raw: feature,
  };
}

/** Public so it can be unit-tested without hitting Mapbox. */
export function classifyConfidence(feature: {
  relevance?: number;
  place_type?: string[];
  properties?: { match_code?: { exact_match?: boolean } };
}): GeocodeConfidence {
  const rel = feature.relevance ?? 0;
  const types = feature.place_type ?? [];
  const exact = feature.properties?.match_code?.exact_match;

  if (rel >= 0.9 && (exact || types.includes("address"))) return "rooftop";
  if (rel >= 0.7 || types.includes("address") || types.includes("street")) return "street";
  if (rel >= 0.5 || types.includes("postcode") || types.includes("place")) return "postcode";
  return "failed";
}

/**
 * Batch helper for the import job — geocodes one customer at a time so a
 * single bad address doesn't poison the batch. Mapbox geocoding is
 * 600 req/min on the free tier; we throttle to ~5/sec to stay safe.
 */
export async function geocodeMany(
  rows: { id: string; address: string }[],
  onProgress?: (done: number, total: number) => void,
): Promise<Record<string, GeocodeResult>> {
  const out: Record<string, GeocodeResult> = {};
  let done = 0;
  for (const r of rows) {
    out[r.id] = await geocodeAddress(r.address);
    done++;
    onProgress?.(done, rows.length);
    await new Promise((res) => setTimeout(res, 200)); // ~5/sec
  }
  return out;
}

File: grasmere_routes/geocode.py
"""Mapbox geocoder bounded to UK, with a confidence band.

Customers below 'street' confidence go to the import review queue —
routing them blindly would put pins in the wrong village.
"""

from __future__ import annotations

import os
import time
from dataclasses import dataclass
from typing import Iterable, Literal

import httpx

GeocodeConfidence = Literal["rooftop", "street", "postcode", "failed"]

MAPBOX_GEOCODE_URL = "https://api.mapbox.com/geocoding/v5/mapbox.places"


@dataclass(frozen=True)
class GeocodeResult:
    lat: float | None
    lng: float | None
    confidence: GeocodeConfidence
    matched_address: str | None
    raw: dict | None


def classify_confidence(feature: dict) -> GeocodeConfidence:
    rel = feature.get("relevance", 0)
    types = feature.get("place_type", []) or []
    exact = (feature.get("properties") or {}).get("match_code", {}).get("exact_match")
    if rel >= 0.9 and (exact or "address" in types):
        return "rooftop"
    if rel >= 0.7 or "address" in types or "street" in types:
        return "street"
    if rel >= 0.5 or "postcode" in types or "place" in types:
        return "postcode"
    return "failed"


def geocode_address(address: str, *, token: str | None = None) -> GeocodeResult:
    token = token or os.environ.get("MAPBOX_TOKEN")
    if not token:
        return GeocodeResult(None, None, "failed", None, {"error": "MAPBOX_TOKEN not set"})
    addr = (address or "").strip()
    if not addr:
        return GeocodeResult(None, None, "failed", None, None)

    from urllib.parse import quote

    params = {
        "access_token": token,
        "country": "gb",
        "limit": 1,
        "types": "address,poi,place,postcode",
        "autocomplete": "false",
    }
    url = f"{MAPBOX_GEOCODE_URL}/{quote(addr, safe='')}.json"
    try:
        resp = httpx.get(url, params=params, timeout=10.0)
        resp.raise_for_status()
        body = resp.json()
    except Exception as e:  # noqa: BLE001
        return GeocodeResult(None, None, "failed", None, {"error": str(e)})

    feats = body.get("features") or []
    if not feats:
        return GeocodeResult(None, None, "failed", None, body)
    f = feats[0]
    center = f.get("center") or [None, None]
    return GeocodeResult(
        lat=center[1],
        lng=center[0],
        confidence=classify_confidence(f),
        matched_address=f.get("place_name"),
        raw=f,
    )


def geocode_many(
    rows: Iterable[tuple[str, str]],
    *,
    on_progress=None,
) -> dict[str, GeocodeResult]:
    out: dict[str, GeocodeResult] = {}
    items = list(rows)
    for i, (cid, addr) in enumerate(items, start=1):
        out[cid] = geocode_address(addr)
        if on_progress:
            on_progress(i, len(items))
        time.sleep(0.2)  # ~5 req/s well under Mapbox free-tier limit
    return out
